diff -up mesa-20.3.0/src/gallium/winsys/radeon/drm/radeon_drm_winsys.c.dma mesa-20.3.0/src/gallium/winsys/radeon/drm/radeon_drm_winsys.c
--- mesa-20.3.0/src/gallium/winsys/radeon/drm/radeon_drm_winsys.c.dma	2020-12-04 03:59:53.578230900 +1000
+++ mesa-20.3.0/src/gallium/winsys/radeon/drm/radeon_drm_winsys.c	2020-12-07 10:01:07.602269090 +1000
@@ -530,6 +530,8 @@ static bool do_winsys_init(struct radeon
       }
    }
 
+   ws->info.num_se = ws->info.max_se;
+
    radeon_get_drm_value(ws->fd, RADEON_INFO_MAX_SH_PER_SE, NULL,
                         &ws->info.max_sh_per_se);
    if (ws->gen == DRV_SI) {
