diff -up mesa-20080311/src/mesa/drivers/dri/intel/intel_screen.c.disable-tex-offset mesa-20080311/src/mesa/drivers/dri/intel/intel_screen.c
--- mesa-20080311/src/mesa/drivers/dri/intel/intel_screen.c.disable-tex-offset	2008-03-11 20:05:30.000000000 -0400
+++ mesa-20080311/src/mesa/drivers/dri/intel/intel_screen.c	2008-03-11 20:04:08.000000000 -0400
@@ -384,10 +384,12 @@ intelHandleBufferAttach(__DRIdrawablePri
    intel_renderbuffer_set_region(rb, region);
 }
 
+#if 0
 static const __DRItexOffsetExtension intelTexOffsetExtension = {
    { __DRI_TEX_OFFSET },
    intelSetTexOffset,
 };
+#endif
 
 static const __DRItexBufferExtension intelTexBufferExtension = {
     { __DRI_TEX_BUFFER, __DRI_TEX_BUFFER_VERSION },
@@ -400,7 +402,9 @@ static const __DRIextension *intelExtens
     &driSwapControlExtension.base,
     &driFrameTrackingExtension.base,
     &driMediaStreamCounterExtension.base,
+#if 0
     &intelTexOffsetExtension.base,
+#endif
     &intelTexBufferExtension.base,
     NULL
 };
