diff -up mesa-20081220/src/mesa/drivers/dri/intel/intel_context.h.dave mesa-20081220/src/mesa/drivers/dri/intel/intel_context.h
--- mesa-20081220/src/mesa/drivers/dri/intel/intel_context.h.dave	2008-12-21 08:46:51.000000000 +1000
+++ mesa-20081220/src/mesa/drivers/dri/intel/intel_context.h	2008-12-21 08:47:04.000000000 +1000
@@ -268,6 +268,7 @@ struct intel_context
    __DRIdrawablePrivate *driReadDrawable;
    __DRIscreenPrivate *driScreen;
    intelScreenPrivate *intelScreen;
+#define drm_i915_sarea _drm_i915_sarea
    volatile struct drm_i915_sarea *sarea;
 
    GLuint lastStamp;
diff -up mesa-20081220/src/mesa/drivers/dri/intel/intel_screen.h.dave mesa-20081220/src/mesa/drivers/dri/intel/intel_screen.h
--- mesa-20081220/src/mesa/drivers/dri/intel/intel_screen.h.dave	2008-12-21 09:09:05.000000000 +1000
+++ mesa-20081220/src/mesa/drivers/dri/intel/intel_screen.h	2008-12-21 09:08:35.000000000 +1000
@@ -68,6 +68,7 @@ typedef struct
 
    __DRIscreenPrivate *driScrnPriv;
 
+#define drm_i915_sarea _drm_i915_sarea
    volatile struct drm_i915_sarea *sarea;
 
    int drmMinor;
