diff -up mesa-20070725/include/GL/glu.h.jx mesa-20070725/include/GL/glu.h
--- mesa-20070725/include/GL/glu.h.jx	2007-07-25 09:36:03.000000000 -0400
+++ mesa-20070725/include/GL/glu.h	2007-07-25 10:27:49.000000000 -0400
@@ -44,19 +44,6 @@
 #define GLAPIENTRYP GLAPIENTRY *
 #endif
 
-#ifdef GLAPI
-#undef GLAPI
-#endif
-
-#  if (defined(_MSC_VER) || defined(__MINGW32__)) && defined(BUILD_GLU32)
-#    define GLAPI __declspec(dllexport)
-#  elif (defined(_MSC_VER) || defined(__MINGW32__)) && defined(_DLL) /* tag specifying we're building for DLL runtime support */
-#    define GLAPI __declspec(dllimport)
-#  else /* for use with static link lib build of Win32 edition only */
-#    define GLAPI extern
-#  endif /* _STATIC_MESA support */
-
-
 #ifndef GLAPI
 #define GLAPI
 #endif
