diff -up mesa-20071127/src/mesa/drivers/dri/i915/intel_tex.c.jx mesa-20071127/src/mesa/drivers/dri/i915/intel_tex.c
--- mesa-20071127/src/mesa/drivers/dri/i915/intel_tex.c.jx	2007-11-27 14:33:54.000000000 -0500
+++ mesa-20071127/src/mesa/drivers/dri/i915/intel_tex.c	2008-02-11 15:49:16.000000000 -0500
@@ -98,7 +98,7 @@ do_memcpy(void *dest, const void *src, s
 }
 
 
-#if DO_DEBUG
+#if defined(DO_DEBUG) && !defined(__ia64__)
 
 #ifndef __x86_64__
 static unsigned
@@ -183,7 +183,7 @@ intelInitTextureFuncs(struct dd_function
    functions->UpdateTexturePalette = 0;
    functions->IsTextureResident = intelIsTextureResident;
 
-#if DO_DEBUG
+#if defined(DO_DEBUG) && !defined(__ia64__)
    if (INTEL_DEBUG & DEBUG_BUFMGR)
       functions->TextureMemCpy = timed_memcpy;
    else
