diff -up mesa-20120716/src/mesa/Makefile.am.dave mesa-20120716/src/mesa/Makefile.am
--- mesa-20120716/src/mesa/Makefile.am.dave	2012-07-17 02:16:46.923268668 -0400
+++ mesa-20120716/src/mesa/Makefile.am	2012-07-17 02:17:09.106991335 -0400
@@ -144,8 +144,7 @@ CLEANFILES += libmesa.a libmesagallium.a
 
 pkgconfigdir = $(libdir)/pkgconfig
 
-if HAVE_OSMESA_DRIVER
-pkgconfig_DATA = osmesa.pc
-else
 pkgconfig_DATA = gl.pc
+if HAVE_OSMESA_DRIVER
+pkgconfig_DATA += osmesa.pc
 endif
