--- Mesa-6.5.2/src/mesa/glapi/dispatch.h.jx	2006-12-12 18:02:22.000000000 -0500
+++ Mesa-6.5.2/src/mesa/glapi/dispatch.h	2006-12-12 18:03:04.000000000 -0500
@@ -3840,4 +3840,9 @@
 
 #endif /* !defined(IN_DRI_DRIVER) */
 
+/* backwards compatibility for X server 1.1.1 */
+#define CALL_IsTextureEXT(a, b) CALL_IsTexture(a, b)
+#define CALL_GenTexturesEXT(a, b) CALL_GenTextures(a, b)
+#define CALL_AreTexturesResidentEXT(a, b) CALL_AreTexturesResident(a, b)
+
 #endif /* !defined( _DISPATCH_H_ ) */
