diff -up Mesa-7.3/src/mesa/drivers/dri/i965/brw_context.c.jx Mesa-7.3/src/mesa/drivers/dri/i965/brw_context.c
--- Mesa-7.3/src/mesa/drivers/dri/i965/brw_context.c.jx	2009-01-21 10:55:47.000000000 -0500
+++ Mesa-7.3/src/mesa/drivers/dri/i965/brw_context.c	2009-02-09 17:55:02.000000000 -0500
@@ -138,10 +138,10 @@ GLboolean brwCreateContext( const __GLco
    /* Advertise the full hardware capabilities.  The new memory
     * manager should cope much better with overload situations:
     */
-   ctx->Const.MaxTextureLevels = 12;
+   ctx->Const.MaxTextureLevels = 13;
    ctx->Const.Max3DTextureLevels = 9;
    ctx->Const.MaxCubeTextureLevels = 12;
-   ctx->Const.MaxTextureRectSize = (1<<11);
+   ctx->Const.MaxTextureRectSize = (1<<12);
    
    /* if conformance mode is set, swrast can handle any size AA point */
    ctx->Const.MaxPointSizeAA = 255.0;
