diff -up Mesa-7.1pre/configs/default.da Mesa-7.1pre/configs/default
--- Mesa-7.1pre/configs/default.da	2007-11-01 16:24:30.000000000 +1000
+++ Mesa-7.1pre/configs/default	2007-11-01 16:24:37.000000000 +1000
@@ -60,7 +60,7 @@ GLW_SOURCES = GLwDrawA.c
 
 # Directories to build
 LIB_DIR = lib
-SRC_DIRS = mesa glu glut/glx glw
+SRC_DIRS = mesa glu glut/glx
 GLU_DIRS = sgi
 DRIVER_DIRS = x11 osmesa
 # Which subdirs under $(TOP)/progs/ to enter:
diff -up Mesa-7.1pre/configs/linux-indirect.da Mesa-7.1pre/configs/linux-indirect
--- Mesa-7.1pre/configs/linux-indirect.da	2007-11-01 16:23:32.000000000 +1000
+++ Mesa-7.1pre/configs/linux-indirect	2007-11-01 16:23:39.000000000 +1000
@@ -47,7 +47,7 @@ GL_LIB_DEPS   = $(EXTRA_LIB_PATH) -lX11 
 
 
 # Directories
-SRC_DIRS = glx/x11 glu glut/glx glw
+SRC_DIRS = glx/x11 glu
 DRIVER_DIRS =
 PROGRAM_DIRS = 
 WINDOW_SYSTEM=dri
diff -up Mesa-7.1pre/configs/linux-dri.da Mesa-7.1pre/configs/linux-dri
--- Mesa-7.1pre/configs/linux-dri.da	2007-11-01 16:24:00.000000000 +1000
+++ Mesa-7.1pre/configs/linux-dri	2007-11-01 16:24:12.000000000 +1000
@@ -60,10 +60,10 @@ USING_EGL=0
 
 # Directories
 ifeq ($(USING_EGL), 1)
-SRC_DIRS = egl glx/x11 mesa glu glut/glx glw
+SRC_DIRS = egl glx/x11 mesa glu glut/glx
 PROGRAM_DIRS = egl xdemos
 else
-SRC_DIRS = glx/x11 mesa glu glut/glx glw
+SRC_DIRS = glx/x11 mesa glu glut/glx
 PROGRAM_DIRS = xdemos
 endif
 
