--- mesa-20080331/configure.ac.BAD	2008-04-12 11:37:32.000000000 -0500
+++ mesa-20080331/configure.ac	2008-04-12 11:39:37.000000000 -0500
@@ -512,6 +512,12 @@
                 DRI_DIRS="mach64 r128 r200 r300 radeon tdfx"
             fi
             ;;
+        sparc*)
+            # Build only the drivers for cards that exist on sparc`
+            if test "x$DRI_DIRS" = x; then
+                DRI_DIRS="mach64 r128 r200 r300 radeon ffb"
+            fi
+            ;;
         esac
         ;;
     freebsd*)
