--- Mesa-6.5.2/src/glx/x11/glxext.c.jx	2006-11-15 09:55:47.000000000 -0500
+++ Mesa-6.5.2/src/glx/x11/glxext.c	2007-01-16 10:48:15.000000000 -0500
@@ -666,6 +666,8 @@
 	if ( do_delete && (m->visualID != 0) ) {
 	    do_delete = GL_FALSE;
 
+	    if (m->visualRating != GLX_NON_CONFORMANT_VISUAL_EXT ||
+		getenv("LIBGL_DEBUG"))
 	    fprintf(stderr, "libGL warning: 3D driver claims to not support "
 		    "visual 0x%02x\n", m->visualID);
 	}
