diff -up mesa-20070725/src/mesa/drivers/dri/Makefile.template.jx mesa-20070725/src/mesa/drivers/dri/Makefile.template
--- mesa-20070725/src/mesa/drivers/dri/Makefile.template.jx	2007-07-25 09:36:04.000000000 -0400
+++ mesa-20070725/src/mesa/drivers/dri/Makefile.template	2007-07-26 10:09:54.000000000 -0400
@@ -70,7 +70,7 @@ SHARED_INCLUDES = \
 
 ##### TARGETS #####
 
-default: depend symlinks $(LIBNAME) $(TOP)/$(LIB_DIR)/$(LIBNAME)
+default: symlinks depend $(LIBNAME) $(TOP)/$(LIB_DIR)/$(LIBNAME)
 
 
 $(LIBNAME): $(OBJECTS) $(MESA_MODULES) $(WINOBJ) Makefile $(TOP)/src/mesa/drivers/dri/Makefile.template
