diff -up mesa-20080331/src/glx/x11/dri2.c.jx mesa-20080331/src/glx/x11/dri2.c
--- mesa-20080331/src/glx/x11/dri2.c.jx	2008-03-31 16:58:48.000000000 -0400
+++ mesa-20080331/src/glx/x11/dri2.c	2008-04-09 10:15:59.000000000 -0400
@@ -37,7 +37,7 @@
 #include <X11/extensions/extutil.h>
 #include "glheader.h"
 #include "xf86drm.h"
-#include "dri2proto.h"
+#include <X11/extensions/dri2proto.h>
 #include "dri2.h"
 
 static char dri2ExtensionName[] = DRI2_NAME;
