diff -up Mesa-7.11/src/glx/glxextensions.c.jx Mesa-7.11/src/glx/glxextensions.c
--- Mesa-7.11/src/glx/glxextensions.c.jx	2011-07-08 21:37:10.000000000 -0400
+++ Mesa-7.11/src/glx/glxextensions.c	2011-10-28 20:26:49.172259452 -0400
@@ -139,7 +139,7 @@ static const struct extension_info known
    { GLX(SGIX_visual_select_group),    VER(0,0), Y, Y, N, N },
    { GLX(EXT_texture_from_pixmap),     VER(0,0), Y, N, N, N },
 #endif
-   { GLX(INTEL_swap_event),            VER(1,4), Y, Y, N, N },
+   { GLX(INTEL_swap_event),            VER(1,4), Y, N, N, N },
    { NULL }
 };
 
