From c016f329abb3d638442bf0b0f27d8f34ebf54f86 Mon Sep 17 00:00:00 2001
From: Kristian Høgsberg <krh@redhat.com>
Date: Tue, 08  Apr  2008  22:26:24  +0000
Subject: Only convert configs if screen creation was successful.

Thanks to Adam Jackson for pointing it out.
---
diff --git a/progs/demos/Makefile b/progs/demos/Makefile
index 456bd4a..01b76ad 100644
--- a/progs/demos/Makefile
+++ b/progs/demos/Makefile
@@ -13,7 +13,7 @@ OSMESA32_LIBS = -L$(TOP)/$(LIB_DIR) -lglut -lOSMesa32 -lGLU -lGL $(APP_LIB_DEPS)
 
 LIB_DEP = $(TOP)/$(LIB_DIR)/$(GL_LIB_NAME) $(TOP)/$(LIB_DIR)/$(GLU_LIB_NAME) $(TOP)/$(LIB_DIR)/$(GLUT_LIB_NAME)
 
-LIBS = -L$(TOP)/$(LIB_DIR) -l$(GLUT_LIB) -l$(GLU_LIB) -l$(GL_LIB) $(APP_LIB_DEPS)
+LIBS = -L$(TOP)/$(LIB_DIR) -l$(GLUT_LIB) -l$(GLU_LIB) -l$(GL_LIB) $(APP_LIB_DEPS) -lX11  -lXi -lXmu
 
 PROGS = \
 	arbfplight \
diff --git a/src/glx/x11/dri_glx.c b/src/glx/x11/dri_glx.c
index 004af0b..fd10047 100644
--- a/src/glx/x11/dri_glx.c
+++ b/src/glx/x11/dri_glx.c
@@ -786,14 +786,16 @@ CallCreateNewScreen(Display *dpy, int scrn, __GLXscreenConfigs *psc,
 							     & driver_configs,
 							     psc);
 
-				    psc->configs =
-					driConvertConfigs(psc->core,
-							  psc->configs,
-							  driver_configs);
-				    psc->visuals =
-					driConvertConfigs(psc->core,
-							  psc->visuals,
-							  driver_configs);
+				    if (psp) {
+					psc->configs =
+					    driConvertConfigs(psc->core,
+							      psc->configs,
+							      driver_configs);
+					psc->visuals =
+					    driConvertConfigs(psc->core,
+							      psc->visuals,
+							      driver_configs);
+				    }
 				}
 			    }
 			}
--
cgit v0.7.2-37-g538c
