diff -up mesa-20071127/src/mesa/drivers/dri/sis/sis_context.h.broken mesa-20071127/src/mesa/drivers/dri/sis/sis_context.h
--- mesa-20071127/src/mesa/drivers/dri/sis/sis_context.h.broken	2008-02-11 20:22:25.000000000 -0500
+++ mesa-20071127/src/mesa/drivers/dri/sis/sis_context.h	2008-02-11 20:23:38.000000000 -0500
@@ -403,6 +403,8 @@ struct sis_context
 
 #if defined(__i386__) || defined(__amd64__)
 #define MMIO_WMB()	__asm __volatile("" : : : "memory")
+#elif defined (__ia64__)
+#define MMIO_WMB()	__asm __volatile ("mf" : : : "memory")
 #else
 #error platform needs WMB
 #endif
