diff -up mesa-20111114/configure.ac.dma mesa-20111114/configure.ac
--- mesa-20111114/configure.ac.dma	2011-11-14 15:57:16.783162469 +0000
+++ mesa-20111114/configure.ac	2011-11-14 15:58:07.877159762 +0000
@@ -1742,7 +1742,7 @@ if test "x$enable_gallium_llvm" = xyes;
     if test "x$LLVM_CONFIG" != xno; then
 	LLVM_VERSION=`$LLVM_CONFIG --version`
 	LLVM_CFLAGS=`$LLVM_CONFIG --cppflags|sed 's/-DNDEBUG\>//g'`
-	LLVM_LIBS="`$LLVM_CONFIG --libs`"
+	LLVM_LIBS=-lLLVM-`llvm-config --version` -lstdc++
 
 	LLVM_LDFLAGS=`$LLVM_CONFIG --ldflags`
 	DEFINES="$DEFINES -D__STDC_CONSTANT_MACROS"
