diff -up mesa-20100529/src/mesa/drivers/dri/intel/intel_fbo.c.dave mesa-20100529/src/mesa/drivers/dri/intel/intel_fbo.c
--- mesa-20100529/src/mesa/drivers/dri/intel/intel_fbo.c.dave	2010-05-30 18:24:35.000000000 +1000
+++ mesa-20100529/src/mesa/drivers/dri/intel/intel_fbo.c	2010-05-30 18:24:42.000000000 +1000
@@ -558,11 +558,6 @@ intel_render_texture(GLcontext * ctx,
        return;
    }
 
-   DBG("Begin render texture tid %lx tex=%u w=%d h=%d refcount=%d\n",
-       _glthread_GetID(),
-       att->Texture->Name, newImage->Width, newImage->Height,
-       irb->Base.RefCount);
-
    /* point the renderbufer's region to the texture image region */
    if (irb->region != intel_image->mt->region) {
       if (irb->region)
